src/gatsby-node.js

```diff
@@ -167,201 +70,372 @@ const processData = async (data, baseUrl, backlinks, token) => {
-
-  return nodes;
+// Fetch only navigation component node
+const fetchPloneNavigationNode = async (id, token, baseUrl) => {
+  // Fetch from Plone REST API and normalize it to be GraphQL compatible
+  const data = normalizeData(
+    await fetchPlone(`${id}/@navigation`, token, {
+      // TODO: Higher depth results in "conflicting field types in your data"
+      'expand.navigation.depth': 1,
+    }),
+    baseUrl
+  );
+  // Yield navigation node
+  return makeNavigationNode(
+    `${id}/@navigation`,
+    data,
+    data._path.split('@navigation')[0]
+  );
 };
 
-// BFS RECURSIVE ALGORITHM
-const processNodesUsingRecursion = async (
-  baseUrl,
-  token,
-  expansions,
-  showLogs
+// GatsbyJS source plugin for Plone
+exports.sourceNodes = async (
+  { actions, cache, getNode, getNodes, store },
+  { baseUrl, token, searchParams, logLevel }
 ) => {
-  let nodes = [],
-    backlinks = {};
-  const queue = [baseUrl];
-  const seen = {};
-
-  logMessage('Traversing the site and fetching data', showLogs);
-  while (queue.length > 0) {
-    const url = queue.shift();
-    seen[url] = true;
-
-    const itemData = await fetchData(url, token, expansions);
-
-    let children = [];
-    if (itemData.batching) {
-      const allItems = await fetchAllItems(url, token);
-      children = allItems.map(item => item['@id']);
-    } else {
-      if (itemData.items) {
-        children = itemData.items.map(item => item['@id']);
+  const { createNode, deleteNode, setPluginStatus, touchNode } = actions;
+  let state = {},
+    newState = {
+      lastFetched: new Date(),
+    };
+  const logger = logging.getLogger(logging[logLevel] || 100);
+
+  logger.info('Reading plugin status');
+  if (
+    store.getState().status.plugins &&
+    store.getState().status.plugins['gatsby-source-plone']
+  ) {
+    state = store.getState().status.plugins['gatsby-source-plone'];
+  }
+  logger.debug(JSON.stringify(state));
+
+  // Normalize baseUrl into form without ending slash
+  baseUrl = baseUrl.replace(/\/+$/, '');
+
+  logger.info('Fetching all metadata');
+  const plone = normalizeData(
+    await fetchPlone(
+      `${baseUrl}/@search`,
+      token,
+      // Search nodes in path order to ensure parents before their children
+      {
+        metadata_fields: 'modified',
+        sort_on: 'path',
+        sort_order: 'ascending',
+        ...searchParams,
+      }
+    ),
+    baseUrl
+  );
+  // Ensure that items include baseUrl (which may be fetched twice)
+  plone.items.push({ _id: baseUrl });
+
+  // Define shared backlinks container to collect links between nodes
+  const backlinks = new Map();
+
+  if (!state.lastFetched) {
+    logger.info('Creating all nodes');
+    for (const item of plone.items) {
+      for await (const node of ploneNodeGenerator(
+        item._id,
+        token,
+        baseUrl,
+        backlinks
+      )) {
+        logger.info(`Creating node – ${node.id.replace(baseUrl, '') || '/'}`);
+        createNode(node);
```

Comment:
@ajayns This is all there is for clean fresh build. So it is pretty clean. The rest of this `sourceNodes` is only for incremental builds (to figure out what has been changed and do necessary updates / removals).